besskge/scoring.py

```diff
@@ -1368,3 +1368,165 @@ def score_tails(
+        """
+        super(TranS, self).__init__(
+            negative_sample_sharing=negative_sample_sharing, scoring_norm=scoring_norm
+        )
+
+        self.sharding = sharding
+        self.normalize = normalize_entities
+
+        if isinstance(entity_initializer, list):
+            entity_initializer = 2 * entity_initializer
+        # self.entity_embedding[..., :embedding_size] : entity embedding
+        # self.entity_embedding[..., embedding_size:] : tilde entity embedding
+        self.entity_embedding = initialize_entity_embedding(
+            self.sharding, entity_initializer, [embedding_size, embedding_size]
+        )
+        if isinstance(relation_initializer, list):
+            relation_initializer = 3 * relation_initializer
+        # self.relation_embedding[..., :embedding_size] : relation embedding
+        # self.relation_embedding[..., :embedding_size] : bar relation embedding
+        # self.relation_embedding[..., :embedding_size] : hat relation embedding
+        self.relation_embedding = initialize_relation_embedding(
+            n_relation_type,
+            inverse_relations,
+            relation_initializer,
+            [embedding_size, embedding_size, embedding_size],
+        )
+        assert (
+            self.entity_embedding.shape[-1] / 2
+            == self.relation_embedding.shape[-1] / 3
+            == embedding_size
+        ), (
+            "Trans requires `2*embedding_size` embedding parameters for each entity"
+            " and `3*embedding_size` embedding parameters for each relation"
+        )
+        self.embedding_size = embedding_size
+        self.register_buffer(
+            "offset", torch.tensor([offset], dtype=self.entity_embedding.dtype)
+        )
+
+    # docstr-coverage: inherited
+    def score_triple(
+        self,
+        head_emb: torch.Tensor,
+        relation_id: torch.Tensor,
+        tail_emb: torch.Tensor,
+    ) -> torch.Tensor:
+        relation_emb = torch.index_select(
+            self.relation_embedding, index=relation_id, dim=0
+        )
+        r, r_bar, r_hat = torch.split(relation_emb, self.embedding_size, dim=-1)
+        head_emb_main, head_emb_tilde = torch.split(head_emb, self.embedding_size, dim=-1)
+        tail_emb_main, tail_emb_tilde = torch.split(tail_emb, self.embedding_size, dim=-1)
+        if self.normalize:
+            head_emb_main = torch.nn.functional.normalize(head_emb_main, p=2, dim=-1)
+            tail_emb_main = torch.nn.functional.normalize(tail_emb_main, p=2, dim=-1)
+            head_emb_tilde = torch.nn.functional.normalize(head_emb_tilde, p=2, dim=-1)
+            tail_emb_tilde = torch.nn.functional.normalize(tail_emb_tilde, p=2, dim=-1)
+        R_h = head_emb_main * (tail_emb_tilde + self.offset)
+        R_t = tail_emb_main * (head_emb_tilde + self.offset)
+        R_r = r_bar * head_emb_main + r + r_hat * tail_emb_main
+        return -self.reduce_embedding(R_h - R_t + R_r)
```

Comment:
This is perfectly correct, but we could maybe save a couple of tensor multiplications by refactoring the scoring function as
`head_emb_main * (tail_emb_tilde + r_bar + self.offset) - tail_emb_main * (head_emb_tilde - r_hat + self.offset) + r`